### ISSUE TITLE
Duplicate column names after join with `[==col]` pattern

### ISSUE THREAD

**Issue by eitsupi** [PRQL/prql#1335]
First, I am excited to actually be able to run queries by PRQL on the playground. It's really awesome!

I have been playing with this wonderful feature and have gotten different results than my intuition.

This is the query.

```elm
from albums
join side:left artists [==artist_id]
take 10
```

This is the result.

```sql
SELECT
  albums.*,
  artists.*
FROM
  albums
  LEFT JOIN artists ON albums.artist_id = artists.artist_id
LIMIT
  10

-- Generated by PRQL compiler version 0.3.1 (https://prql-lang.org)
```

![image](https://user-images.githubusercontent.com/50911393/209457190-98ef76d7-12cf-482c-a117-5b914a4d75e8.png)

I would like a result with only one `artist_id`.

At least with DuckDB it seems that the following query can do this, but I don't know if it can be done with a general translation to SQL......

```sql
FROM
  albums
  LEFT JOIN artists USING(artist_id)
LIMIT
  10
```

**Comment by max-sixty** [comment]
Yes, this is one tradeoff of not using `USING` — we don't collapse the duplicate columns.

We had some discussion on this in https://github.com/PRQL/prql/issues/1206 & https://github.com/PRQL/prql/issues/1138 (and possibly others).

In isolation I'm definitely a fan of `USING`, it's really helpful, especially when exploring data on a schema where columns are similarly named across tables.

---

But atm it makes the compiler a bit more complicated. 

And it doesn't compose well with other conditions; e.g. what should this give:

```
from albums
join side:left artists [==artist_id, album.title==artist.name]
take 10
```

since in SQL if we use `USING`, we can _only_ use `USING`. So in that case we could
- Change to use `ON` for `artist_id` rather than `USING` — though then adding `album.title==artist.name` changes the join condition of the unchanged term `==artist_id`, which is not orthogonal
- Raise an error, requiring `album.artist_id==artist.artist_id`

---

Overall I think `JOIN`s are one area we haven't really upgraded over SQL. My prior is that doing them well requires some semantic model (i.e. would be great to have a map of which columns the tables are designed to join on), or something more ambitious like https://github.com/PRQL/prql/issues/716

So to the extent we can solve this easily, I'm v keen. But it might be that we have to live in this imperfect world for a bit while we figure out an excellent design.

**Comment by aljazerzen** [comment]
Just to add this: the two `artist_id` columns are not exactly the same - because of your left join, the output may contain albums whose artists were not found. In this case "album.artist_id" is set, but "artist.artist_id" is null.

**Comment by max-sixty** [comment]
> Just to add this: the two `artist_id` columns are not exactly the same - because of your left join, the output may contain albums whose artists were not found. In this case "album.artist_id" is set, but "artist.artist_id" is null.

@aljazerzen is completely correct, I should have noted this above. They're only the same column if `side:inner`.

(there's a reasonable case that we should default to `inner` for the `==foo` join. But it's unclear and I mostly put it in the "Joins are OK but not a big upgrade yet" bucket)

**Comment by eitsupi** [comment]
Thank you for the detailed explanation.
And sorry I did not realize that there have been similar issues in the past.

My wish is for it to work like dbplyr, but I understand that this is obviously possible because dbplyr is aware of the tables' schema (column names), which is not possible because prql-compiler is currently not aware of the tables' schema.

``` r
df_artists <- readr::read_csv("https://raw.githubusercontent.com/PRQL/prql/d3a646be8098b926d3dd5870fcc1ef22a26b7db7/prql-compiler/tests/integration/data/chinook/artists.csv", show_col_types = FALSE)
df_albums <- readr::read_csv("https://raw.githubusercontent.com/PRQL/prql/d3a646be8098b926d3dd5870fcc1ef22a26b7db7/prql-compiler/tests/integration/data/chinook/albums.csv", show_col_types = FALSE)

con <- DBI::dbConnect(duckdb::duckdb(), ":memory:")

duckdb::duckdb_register(con, "artists", df_artists)
duckdb::duckdb_register(con, "albums", df_albums)

dplyr::tbl(con, "albums") |>
  dplyr::left_join(dplyr::tbl(con, "artists"), by = "artist_id") |>
  head(10) |>
  dplyr::show_query()
#> <SQL>
#> SELECT *
#> FROM (
#>   SELECT album_id, title, LHS.artist_id AS artist_id, "name"
#>   FROM albums AS LHS
#>   LEFT JOIN artists AS RHS
#>     ON (LHS.artist_id = RHS.artist_id)
#> ) q01
#> LIMIT 10
```

<sup>Created on 2022-12-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by max-sixty** [comment]
Yes, very much agree @eitsupi .

Do you have thoughts on what a good design would be? Would you want to default to `INNER JOIN`? To `USING`?

Because while I recognize we have lots of tradeoffs, and maybe are in the best place given all of them — it does seem unfriendly that this doesn't generate a nice result:

```elm
from albums
join artists [==artist_id]
take 10
```

**Comment by eitsupi** [comment]
> Would you want to default to `INNER JOIN`? To `USING`?

I like to convert the `[==col]` pattern to `USING` and `[==col, table1.foo==table2.bar]` to error.

Ideally, there would be an option to pass the table schema to the compiler, and if that option is used, it would remove duplicate key columns and automatically correct column names like dbplyr does, but perhaps it is not desirable to have the behavior change with or without the option?
(Since you are working on #1329, perhaps such an idea has already been considered)

**Comment by aljazerzen** [comment]
Why do you want `USING`?

I've changed the compiler to not use it because it would sometimes produce three copies of `artist_id` and because of inconsistent behavior across databases.

I really dislike changing the behavior of join depending on whether we have database schema or not, but we could add it a `dedup_id:false` named arg.

**Comment by eitsupi** [comment]
> Why do you want `USING`?
>
> I've changed the compiler to not use it because it would sometimes produce three copies of `artist_id` and because of inconsistent behavior across databases.

Thanks for explaining all the details!

I just wanted a result without duplicate columns here, not that I want to use `USING`.
I simply knew that in DuckDB SQL we can use `USING` to get such a result.

I don't have strong opinion on changing the behavior of the compiler, and if the current behavior is what was intended, that is still fine.
If I knew how this works I would just write a query like the following.

```elm
from albums
join side:left artists [==artist_id]
select [
  albums.*,
  artists.name
]
take 10
```

However, if it were not necessary to write `select` for this query, or to write `albums.` or `artists.` in the `select`, that would be wonderful I think.

**Comment by max-sixty** [comment]
I'm aligned with @eitsupi's perspective... Let me give a quick summary of my logic for `USING` and then some problems with relying on it:

### Case for `USING` / equi-joins

It's really clear what's happening here:

```elm
from albums
join artists [==artist_id]
join tracks [==album_id]
```

If the tables' schema is designed with this intention, it's ideal. Indeed — it reduces the need for a semantic model which encapsulates the join conditions (the only advantage from a full semantic model would be not having to list the columns).

...compared with...

```elm
from albums
join artists [==artist_id]
join tracks [artists.album_id==tracks.album_id]

 # As per @eitsupi 's comment, we need to list the column if we don't want duplicate columns
select [  
  albums.*,
  artists.artist_name,  # List out all the columns from `artists` here!
  tracks.track_name,  # And again do the list for `tracks`...
]
```



### Case against

- Not all DBs support `USING`. A quick glance suggests this is only `mssql` though? So it _might_ be possible to just raise an error on `[==foo]` for that dialect (or stop supporting the dialect, as a drastic move...)
- It requires us to default to Inner joins — (edit: we do this already, it's a good default I think)
- It's a bit more magic. I need to spend more time with this part of the compiler to understand how much it complicates how we represent columns — IIUC there isn't really a way of representing "all columns from `albums` and `artists`, _except for the columns which we joined on_".

**Comment by aljazerzen** [comment]
If I expand what I proposed before:

We can add a named argument to join. 
```
dedup_id<bool>
# when true, all columns from right relation that are mentioned in `filter` param
# will be omitted from the output relation.
```
The default could even be set to `true`, for the convenience.

This should ideally be implemented without `USING` because:
- it's not supported by all databases,
- it's quite inconsistent when you use anything else than `*` in SELECT

Example:
```elm
from albums
join artists [==artist_id]
```
```sql
SELECT albums.*, artists.name
FROM albums
JOIN artists ON albums.artist_id = artists.artist_id
```
Note that for this query, the compiled needed the knowledge of all columns of artists table. This problem also shows up in #1329 where I've outlined how to solve it.

**Comment by max-sixty** [comment]
I realize we've discussed this a lot, so I don't mean to keep pestering here. 

But I have really liked using `USING` when I write SQL — I even design my tables around it. And having the restriction of the compiler needing to know the underlying tables is fairly restrictive. So one follow-up:

> it's quite inconsistent when you use anything else than `*` in SELECT

What's a good case of this? Maybe if we can reconcile this then we can synthesize our various perspectives.

> it's not supported by all databases,

I agree and weigh this quite highly. But IIUC it's _only_ MSSQL, which we could just raise an error on any use of `[==foo]`; we could do this fairly late in the compiler pipeline such that we don't need dialect information early.

**Comment by aljazerzen** [comment]
The problem was with wildcards:
```sql
SELECT * FROM a JOIN b USING (some_id)  -- all is well
SELECT a.*, b.* FROM a JOIN b USING (some_id)  -- both a.some_id and b.some_id are returned
SELECT a.*, b.*, some_id FROM a JOIN b USING (some_id)  -- a.some_id, b.some_id and some_id
```
I know - last example does not make sense, but with previous impl of join it sometimes got emitted and caused very unexpected results.

---

Why should we care about compiler emitting USING? If we go with my proposal, PRQL side is just as ergonomic as USING is SQL, but emitted SQL uses ON.

**Comment by max-sixty** [comment]
> ```sql
> SELECT a.*, b.* FROM a JOIN b USING (some_id)  -- both a.some_id and b.some_id are returned
> ```

Yes, I definitely empathize that it makes the column & namespace transformations more complicated — we get an implicit change in what's available from the namespace:
- From: all the columns in `a` & all the columns in `b`
- To: all the columns in `a` & all the columns in `b`, except for `some_id` from both `a` & `b`, which changes to a new non-namespaced column (!)

And ofc I recognize you've done all the work here.

> Why should we care about compiler emitting USING? If we go with my proposal, PRQL side is just as ergonomic as USING is SQL, but emitted SQL uses ON.

Yes but it requires knowing the table definitions. I think it's going to be great for the developer experience to have those available when typing a query. But if table definitions are _required for the compilation_, we should put quite a lot of cost on that. One of PRQL's advantages over comparable langs is that it doesn't require much context to compile a query, so it's easy to try, easy to observe, easy to test, compatible with any DB, etc etc

OTOH in exchange for that simplicity, we lose things like a semantic model and automatic joins. Joins are the one area where a semantic model would be a big advantage, so I guess this is why we're at this tradeoff now. 

And maybe going from "SQL->PRQL" to "SQL+Table_def->PRQL" isn't as bad as I fear.

But if we can avoid requiring table definitions for the compilation, then I would strongly vote for us to really try.

**Comment by aljazerzen** [comment]
Ah, that's true and sad. Most dialects don't support EXCLUDE so we cannot replicate USING with ON.

Well in this case, I'd prefer to detect when to use USING automatically and not relying on self equality operator. Which would means that this:
```elm
from artists
join albums [artists.artist_id == albums.artist_id] dedup_id:false
select -[albums.name]
```
... can also compile to:
```sql
SELECT * FROM artists JOIN albums USING(artist_id)
```
... for dialects that do support it.
